enhancements/20231108-elemental-cloudinit-crd.md

```diff
@@ -0,0 +1,376 @@
+  applied at boot.
+
+* Stanza/block-level editing. The CloudInit resource's `.spec.contents`
+  field will fully overwrite the file that is referred to in the resource's
+  `.spec.filename` field.
+
+## Proposal
+
+### User Stories
+
+Harvester users may customize their Harvester nodes to the fullest extent
+that the [Elemental toolkit](
+https://rancher.github.io/elemental-toolkit/docs/reference/cloud_init/) allows.
+
+Because of this, there is a very broad and nuanced set of customizations a Harvester
+user can describe in a cloud-init-style file.
+
+#### Add/rotate rancher user authorized keys (API)
+
+For example, a user could distribute and/or rotate the `rancher` user's authorized
+keys by creating a CloudInit resource like so:
+
+```yaml
+apiVersion: node.harvesterhci.io/v1beta1
+kind: CloudInit
+metadata:
+  name: ssh-access
+spec:
+  matchSelector: {}
+  filename: 99_ssh.yaml
+  contents: |
+    stages:
+      network:
+        - authorized_keys:
+            rancher:
+              - ssh-ed25519 AAAA...
+```
+
+Then, the user could monitor the ssh-access CloudInit resource to know the status
+of its rollout to the Harvester cluster:
+
+```
+$ kubectl describe cloudinit ssh-access
+# Metadata and Spec omitted for brevity...
+Status:
+  Rollouts:
+    Conditions:
+      Last Transition Time:  2023-10-24T21:15:02Z
+      Message:               
+      Reason:                CloudInitApplicable
+      Status:                True
+      Type:                  Applicable
+      Last Transition Time:  2023-10-24T21:09:28Z
+      Message:               Local file checksum is the same as the CloudInit checksum
+      Reason:                CloudInitChecksumMatch
+      Status:                False
+      Type:                  OutOfSync
+      Last Transition Time:  2023-10-24T21:09:28Z
+      Message:               99_ssh.yaml is present under /oem
+      Reason:                CloudInitPresentOnDisk
+      Status:                True
+      Type:                  Present
+    Node Name:               harv1
+```
+
+In the case of this single-node cluster, the user can see that the
+cloud-init-style file has been copied down to `/oem/99_ssh.yaml` on
+node `harv1`.
+
+They can also see that the file is up-to-date since the rollout says
+it is not `OutOfSync`.
+
+The user probably already knows that it is `Applicable` to `harv` because
+the `ssh-access` CloudInit resource spec says `matchSelector: {}`, which is
+a label set that will match every node. That said, if they updated the
+`matchSelector` down the line, then the `Applicable` condition may help
+them troubleshoot why a node has or does not have a cloud-init-style file
+copied down to it.
+
+The user can now reboot the Harvester node so that their cloud-init-style
+file may be applied at boot time.
+
+#### Add/rotate rancher user authorized keys (UI/Dashboard)
+
+Alternatively, instead of creating the `ssh-access` resource via kubectl,
```

Comment:
We have a page, 'SSH Keys', under Advanced tab. Is it related in some way with the new CloudInit functionality ? If yes, we may consider to put the new page along with the SSH key page.

Reply:
No, it's a sad naming coincidence. This is for "Elemental cloud init" for customizing the host Harvester OS, but the cloud-init that many people are familiar with is for customizing a virtual machine as it boots up. The SSH Keys page is for storing SSH public keys that can be injected into the VM as part of the VM-based cloud-init and not the Elemental cloud-init.